File: A0/dot.c
#include "mpi.h"
#include <stdio.h>
#include <stdlib.h>

double parallelDot(double* localx,double* localy,int n);
double serialDot(double* x, double* y,int n);

int main(int argc, char *argv[])  {
    int numtasks, rank, dest, source, rc, count, tag=1;  
    char inmsg, outmsg='x';
    MPI_Status Stat;   // required variable for receive routines

    MPI_Init(&argc,&argv);
    MPI_Comm_size(MPI_COMM_WORLD, &numtasks);
    MPI_Comm_rank(MPI_COMM_WORLD, &rank);

    if(rank==0){
        printf("Testing Parallel Dot Product...\n");
    }
    //root creates arrays and then scatters to everyone
    int n=1000;
    int i;
    double* x;
    double* y;
    //have main allocate and fill the matrices
    if (rank == 0) {
        x=(double*) malloc(n*sizeof(double));
        y=(double*) malloc(n*sizeof(double));
        for(i=0;i<n;i++){
            x[i]=1.0;
            y[i]=1.0;
        }
    }
    //create the local matrices for the chuncks
    int localn=n/numtasks;
    double* localx = (double*) malloc(localn*sizeof(double));
    double* localy = (double*) malloc(localn*sizeof(double));

    //scatter to all
    MPI_Scatter(x,localn,MPI_DOUBLE,localx,localn,MPI_DOUBLE,0,MPI_COMM_WORLD);                  
    MPI_Scatter(y,localn,MPI_DOUBLE,localy,localn,MPI_DOUBLE,0,MPI_COMM_WORLD);

    //run the parallel dot product
    //all ranks must call this function, but only root will
    //have the correct final result
    double res = parallelDot(localx,localy,localn);
    
    //print result
    if(rank==0){
        if(res==(double)n){
            printf("Test Passed\n");
        }else{
            printf("Test Failed\n");
        }
    }
       
    //deallocate memory
    if(rank=0){
        free(x);
        free(y);
    }
    free(localx);
    free(localy);
        
    MPI_Finalize();
    
}

//serial version
double serialDot(double* x, double* y,int n){
    double sum=0.0;
    int i;
    for(i=0;i<n;i++){
        sum+=x[i]*y[i];
    }
    return sum;
}

//parallel version (uses serial version for local dot product)
double parallelDot(double* localx,double* localy,int n){
    double dot=0.0;
    double localDot=0.0;
    localDot=serialDot(localx,localy,n);
    MPI_Reduce(&localDot,&dot,1,MPI_DOUBLE,MPI_SUM,0,MPI_COMM_WORLD);
    return dot;
}

File: A0/dot.f90
module myDot
    implicit none
    include 'mpif.h'

    contains

    real(kind=8) function serialDot(x,y) result(sum)
        implicit none
        real(kind=8), dimension(:) :: x,y
        integer :: i

        if(size(x) .ne.  size(y)) then
            sum=0
        else

            do i=1,size(x)
                sum=sum+x(i)*y(i)
            enddo
        endif

    end function serialDot

    real(kind=8) function parallelDot(x,y) result(dot)
        implicit none
        include 'mpif.h'
        real(kind=8) :: mysum
        integer ierr,i
        real(kind=8), dimension(:) :: x,y
        if(size(x).ne.size(y)) then
            dot=0
        else
            mysum = serialDot(x,y)
            call MPI_REDUCE(mysum,dot,1,MPI_DOUBLE_PRECISION,MPI_SUM,0,MPI_COMM_WORLD,ierr)
        endif

    end function parallelDot
end module myDot



program dot
    use myDot
    implicit none
    integer numtasks, rank, dest, source,count, tag, ierr
    integer stat(MPI_STATUS_SIZE)   !required variable for receive routines
    real(kind=8), allocatable, dimension(:) :: x,y,localx,localy
    integer n,localn,i
    real(kind=8) :: res

    call MPI_INIT(ierr)
    call MPI_COMM_RANK(MPI_COMM_WORLD, rank, ierr)
    call MPI_COMM_SIZE(MPI_COMM_WORLD, numtasks, ierr)
    if(rank==0) then
        WRITE(*,'(A)') "Testing Parallel Dot Product..."
    endif

    n=1000
    localn=n/numtasks

    !root allocates the inital arrays
    if(rank==0) then
        allocate(x(n),y(n))
        do i=1,n
            x(i)=1.0
            y(i)=1.0
        enddo
    endif
    !everyone allocates their local arrays
    allocate(localx(localn),localy(localn))

    !scatter
    call MPI_SCATTER(x,localn,MPI_DOUBLE_PRECISION,localx,localn,MPI_DOUBLE_PRECISION,0,MPI_COMM_WORLD,ierr)
    call MPI_SCATTER(y,localn,MPI_DOUBLE_PRECISION,localy,localn,MPI_DOUBLE_PRECISION,0,MPI_COMM_WORLD,ierr)

    !call parallel dot product (includes the mpi reduce)
    res=parallelDot(localx,localy) 
    

    if(rank==0) then
        if(res==n) then
            Write(*,'(A)') "Test Passed"
        else
            Write(*,'(A)') "Test Failed"
        endif
        Write(*,*)
    endif


    if(allocated(x)) deallocate(x)
    if(allocated(y)) deallocate(y)
    if(allocated(localx)) deallocate(localx)
    if(allocated(localy)) deallocate(localy)
    call MPI_FINALIZE(ierr)

end program dot
